contrib/nitiwiki/src/wiki_edit.nit

```diff
@@ -0,0 +1,182 @@
+import markdown
+import md5
+
+intrude import wiki_html
+
+# Page for editing markdown source
+class WikiEditForm
+	super WikiArticle
+
+	# Part of the title before the name of the page
+	var title_prefix: String
+
+	# Markdown content, for previews
+	redef var md
+
+	# Custom HTML code, for forms and links
+	var html: String
+
+	init do content = (md or else "").md_to_html.to_s + html
+
+	redef fun dir_href do return href.dirname
+
+	redef fun tpl_article
+	do
+		var s = super
+		s.title = title_prefix + title
+		return s
+	end
+
+	# Fill and return a new `HttpResponse` with this page content
+	fun to_http_response: HttpResponse
+	do
+		var resp = new HttpResponse(200)
+		resp.body = tpl_page.write_to_string
+		return resp
+	end
+end
+
+# Action to serve edit forms, show previews and apply changes
+class EditAction
+	super Action
+
+	# Full public URL for the root of this wiki
+	var root_url: String
+
+	# Path to the wiki config
+	var config_file_path: String
+
+	# Configuration of the Wiki, loaded once
+	var wiki_config = new WikiConfig(config_file_path) is lazy
+
+	# Path to the root of the wiki
+	var wiki_root: String = config_file_path.dirname is lazy
+
+	# List of acceptable password to apply modifications
+	#
+	# If `null`, no password checks are applied and all modifications are accepted.
+	var passwords: nullable Collection[String]
+
+	# Reload the wiki instance with the latest changes
+	fun wiki: Nitiwiki
+	do
+		var wiki = new Nitiwiki(wiki_config)
+		wiki.parse
+		return wiki
+	end
+
+	redef fun answer(http_request, turi)
+	do
+		var action = http_request.string_arg("action")
+		var markdown = http_request.post_args.get_or_default("content", "")
+
+		var file_path = turi.strip_leading_slash
+		file_path = wiki_root / file_path
+
+		if action == "Submit" then
+			var passwords = passwords
+			var password = http_request.post_args.get_or_null("password")
+			if passwords != null and (password == null or not passwords.has(password.md5)) then
+				# Deny modification
+				var entity = new WikiEditForm(wiki, turi.strip_leading_slash,
+					"Changes rejected: ", "", "<p>Wrong password</p>")
+				return entity.to_http_response
+			end
+
+			# Save markdown source
+			markdown = markdown.replace('\r', "")
+			markdown.write_to_file file_path
+
+			# Update HTML files
+			var wiki = wiki
+			wiki.render
+
+			var link
+			if turi.has_prefix("/pages/") then
+				link = root_url / turi.substring_from(7)
+			else link = root_url / turi
+			link = link.strip_extension(".md") + ".html"
+
+			# Show confirmation
+			var body = """
+<p>Your edits were recorded and the file is updated: <a href="{{{link}}}">{{{link}}}</a></p>
+"""
+			var entity = new WikiEditForm(wiki, turi.strip_leading_slash, "Changes saved: ", "", body)
+			return entity.to_http_response
+		else
+			# Show edit form, and preview when requested
+
+			# When not in a preview, use the local content of the file
+			if action != "Preview" then markdown = file_path.to_path.read_all
```

Comment:
injection of path: It seems that I might be able to read /etc/shadow or .ssh with a correctly engineered turi